File: backend/services/journal.py
"""One append-only record of what happened to a recitation.

Everything backend.log already says is still said there; this is a second,
narrower log, one JSON object per line, built to answer one question fast:
"where did this particular reading go wrong, server side or page side, and at
what point". A reading id ties the two sides together, so a support
conversation is "send me reading abc123" rather than a stretch of prose logs
either side has to be read start to end to place.

Every line has the same shape: {"at", "side": "server"|"page", "kind", plus
whatever that kind of event knows}. `note` writes the server's half, called
from routers/listen.py and services/recitation/ears.py. `note_page` writes the
page's half, called only from routers/journal.py, which is the one door the
browser has onto this file.

Its own logger, and propagate=False, so a line written here does not also land
in backend.log: the two logs answer different questions and mixing them would
make both slower to read.
"""
from __future__ import annotations

import contextvars
import json
import logging
import logging.handlers
import re
import threading
from datetime import datetime
from pathlib import Path
from typing import Any

from backend.config import get_settings

# backend/services/journal.py -> backend/services -> backend -> project root,
# the folder that holds backend/. journal_path is relative to here rather than
# to backend/ the way data_path's paths are: the journal belongs beside
# backend.log, not inside the app's own data folder.
_PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent

_journal = logging.getLogger("recite.journal")
_journal.propagate = False
# Which settings the handler currently open was built from, so a changed
# journal_path (a test pointing it at tmp_path, or a reloaded .env) is picked
# up on the next line instead of writing to wherever the first line went.
_built_from: tuple[str, int, int] | None = None
# Two request threads can both find the handler stale at once; without a
# lock both would add one, and every line after would be written twice.
_configure_lock = threading.Lock()


class JournalConfigError(ValueError):
    """A journal setting is not usable. Raised the first time something is
    actually logged, so a bad .env value fails loud rather than the app
    silently writing nowhere."""


def _ensure_configured() -> None:
    global _built_from
    settings = get_settings()
    if settings.journal_max_kb <= 0:
        raise JournalConfigError("journal_max_kb must be positive")
    if settings.journal_keep <= 0:
        raise JournalConfigError("journal_keep must be positive")
    key = (settings.journal_path, settings.journal_max_kb, settings.journal_keep)
    if key == _built_from:
        return
    # Two worker threads can both see a stale _built_from at once; without
    # this lock both add a handler, and every line after is written twice.
    with _configure_lock:
        if key == _built_from:
            return
        for old in list(_journal.handlers):
            _journal.removeHandler(old)
            old.close()
        path = (_PROJECT_ROOT / settings.journal_path).resolve()
        path.parent.mkdir(parents=True, exist_ok=True)
        handler = logging.handlers.RotatingFileHandler(
            path, maxBytes=settings.journal_max_kb * 1024, backupCount=settings.journal_keep, encoding="utf-8",
        )
        handler.setFormatter(logging.Formatter("%(message)s"))
        _journal.addHandler(handler)
        _journal.setLevel(logging.INFO)
        _built_from = key


# The reading a deep call (ears.py, several layers below the router) is
# journalling for, without threading an id as an argument through every
# function in between. Set once per request in routers/listen.py and read by
# anything that wants to say which reading it was working on. Safe across the
# thread a recitation is heard on: run_in_threadpool copies the current
# context into that thread, so a value set here before the call is visible
# inside it.
reading_id: contextvars.ContextVar[str] = contextvars.ContextVar("reading_id", default="")


# A raw SDK exception carries the request that made it, headers and all,
# which is the API key (see groq_backend.py's own note on this). A bearer
# header or a gsk_-prefixed key is the shape that leaks; both are scrubbed
# before a message or traceback ever reaches the journal file.
_SECRET = re.compile(r"Bearer\s+\S+|gsk_[A-Za-z0-9]+", re.IGNORECASE)
_SCRUB_LIMIT = 300


def scrub(text: str) -> str:
    """A raw error message or traceback, made safe to journal: no bearer
    token or key, and capped so one huge message cannot blow up a line."""
    return _SECRET.sub("[redacted]", text)[:_SCRUB_LIMIT]


def _jsonable(value: Any) -> Any:
    """Anything that cannot go straight into JSON, stringified rather than
    dropped: a half-written line is worse than an ugly field."""
    try:
        json.dumps(value)
        return value
    except TypeError:
        return str(value)


def note(kind: str, **fields: Any) -> None:
    """One line from the server side of the app.

    `reading` is filled from the current reading (see `reading_id`) unless the
    caller names one. Fields that are None or "" are left out rather than
    written, so a line only carries what it actually knows.
    """
    fields.setdefault("reading", reading_id.get())
    line: dict[str, Any] = {
        "at": datetime.now().astimezone().isoformat(timespec="milliseconds"),
        "side": "server",
        "kind": kind,
    }
    for key, value in fields.items():
        if value not in (None, ""):
            line[key] = _jsonable(value)
    _ensure_configured()
    _journal.info(json.dumps(line, ensure_ascii=False))


def note_page(event: dict[str, Any]) -> None:
    """One line the page reported about itself, through POST /api/journal.

    Written exactly as sent: two identical events in one batch make two
    identical lines, on purpose. The page may legitimately repeat an event (a
    retry, a mic gate closing twice), and deciding which repeat was the "real"
    one is not this file's job, it is a question for whoever reads the log.
    """
    line: dict[str, Any] = {
        "at": event.get("at"),
        "side": "page",
        "kind": event.get("kind", ""),
    }
    for key in ("session", "reading", "detail"):
        value = event.get(key)
        if value not in (None, ""):
            line[key] = _jsonable(value)
    _ensure_configured()
    _journal.info(json.dumps(line, ensure_ascii=False))

File: frontend/src/lib/journal.js
/**
 * A trail of what reciting did, sent to the server in small batches.
 *
 * Best-effort only: an event that cannot be sent is dropped, never retried,
 * and nothing here ever throws or waits on reciting itself. Read by
 * lib/recitingSession.js; api.js stays the only place that knows the URL.
 */
import { JOURNAL_PATH, postJournal } from '../api'
import config from '../recite.json'

const FLUSH_MS = config['journal-flush-ms']
const BATCH = config['journal-batch']
const QUEUE_MAX = config['journal-queue-max']

let queue = []
let dropped = 0

/** The first `max` events of `queue`, in order; an exact duplicate is kept, not folded away. */
export const batchOf = (queue, max) => queue.slice(0, max)

const post = (events, beacon) => {
  if (beacon && typeof navigator !== 'undefined' && navigator.sendBeacon) {
    navigator.sendBeacon(JOURNAL_PATH, JSON.stringify({ events }))
    return Promise.resolve()
  }
  return postJournal(events)
}

/** Send whatever is queued, in batches of `journal-batch`. Never throws. */
async function flush(beacon = false) {
  while (queue.length) {
    const batch = batchOf(queue, BATCH)
    queue = queue.slice(batch.length)
    try {
      await post(batch, beacon)
    } catch {
      // A batch that fails to send is not retried; the next tick has fresher events anyway.
    }
  }
}

/**
 * Queue one event. `fields.session` and `fields.reading` are the ids it
 * belongs to; anything else in `fields` becomes `detail`. Never throws.
 */
export function note(kind, fields = {}) {
  try {
    const { session, reading, ...detail } = fields
    const event = { at: new Date().toISOString(), kind }
    // Left out when blank: the server refuses a blank id, and one bad event
    // throws away the whole batch it came in.
    if (session) event.session = session
    if (reading) event.reading = reading
    if (Object.keys(detail).length) event.detail = detail
    if (queue.length >= QUEUE_MAX) {
      dropped += 1
    } else {
      queue.push(event)
      if (dropped && queue.length < QUEUE_MAX) {
        queue.push({ at: new Date().toISOString(), kind: 'journal.dropped', detail: { count: dropped } })
        dropped = 0
      }
    }
    if (queue.length >= BATCH) flush()
  } catch {
    // Journalling a recitation must never be the reason it stops.
  }
}

// Flushed on a timer, and again right before the page is gone: the timer
// alone would lose whatever was queued in the last flush-ms of a session.
// Guarded for a test environment with no window: nothing here should run cold.
if (typeof window !== 'undefined') {
  setInterval(() => flush(false), FLUSH_MS)
  const onHide = () => flush(true)
  window.addEventListener('pagehide', onHide)
  document.addEventListener('visibilitychange', () => {
    if (document.visibilityState === 'hidden') onHide()
  })
}
